File: lib/aura/kernel/runner.rb
# frozen_string_literal: true

require "json"
require "open3"
require "aura"
require "aura/context"
require "aura/memory"
require "aura/kernel/registry"
require "aura/kernel/tool_validator"
require "aura/kernel/execution_engine"
require "aura/ext/lsp/manager"
require "aura/kernel/planner"
require "aura/config_loader"
require_relative "event_emitter"
require_relative "hooks"
require_relative "job"

module Aura
  module Kernel
    class Runner
      include EventEmitter

      attr_reader :hooks, :current_job, :memory, :planner, :project_path, :env_path
      alias workspace_path project_path

      def initialize(project_path, memory: nil)
        @project_path = File.expand_path(project_path)
        @env_path = Aura::PathResolver.environment_path(@project_path)

        @registry = Aura::Kernel::ToolRegistry.new(@env_path)
        @memory = memory || default_memory
        @validator = Aura::Kernel::ToolValidator.new(@project_path, @registry, @memory, env_path: @env_path)
        @lsp_manager = Aura::LSP::Manager.new(@project_path)
        @engine = Aura::Kernel::ExecutionEngine.new(@project_path, env_path: @env_path, lsp_manager: @lsp_manager)
        @context_manager = Aura::Context::Manager.new(@env_path)
        @hooks = Aura::Kernel::Hooks.new
        @planner = Aura::Kernel::Planner.new(@project_path, env_path: @env_path)
        @current_job = nil
        @lock = Mutex.new

        at_exit do
          @memory.store.close
        rescue StandardError
          nil
        end
      end

      # Hot-swap the active SQLite database session to achieve physical amnesia without runner re-instantiation
      def reconnect_session!(session_name)
        ENV["AURA_SESSION_NAME"] = session_name
        if @memory && @memory.respond_to?(:store) && @memory.store
          begin
            @memory.store.close
          rescue StandardError
            nil
          end
        end
        @memory = default_memory
        @validator = Aura::Kernel::ToolValidator.new(@project_path, @registry, @memory, env_path: @env_path)
        @planner = Aura::Kernel::Planner.new(@project_path, env_path: @env_path)
      end

      def load_config
        Aura::ConfigLoader.load(@env_path, safe: true)
      end

      def start_job(metadata = {})
        @lock.synchronize do
          raise "Runner is busy with job #{@current_job.id}" if @current_job && @current_job.status == :running

          @current_job = Aura::Kernel::Job.new(metadata)
          @current_job.start!
        end
        emit(:job_start, @current_job.to_h)
        @current_job
      end

      def end_job(status = :completed, error = nil)
        job = nil
        @lock.synchronize do
          return unless @current_job

          if status == :failed && error
            @current_job.fail!(error)
          else
            @current_job.complete!
          end
          job = @current_job
          @current_job = nil
        end

        emit(:job_end, job.to_h)
        job
      end

      def observe
        @memory.recorder.record_custom("observe", {})
        @memory.metabolize_if_needed
        auto_verify_core_tools
        Aura::Context.assemble(@project_path, @memory, lsp_manager: @lsp_manager)
      end

      def plan(goal = nil, context = nil)
        ctx = context || observe

        payload = { context: ctx, goal: goal }
        @hooks.run(:before_planning, payload)
        ctx = payload[:context]
        goal = payload[:goal]

        res = @planner.plan(ctx, goal)
        @memory.recorder.record_plan(res)
        res
      end

      def plan_stream(goal = nil, context = nil)
        ctx = context || observe

        payload = { context: ctx, goal: goal }
        @hooks.run(:before_planning, payload)
        ctx = payload[:context]
        goal = payload[:goal]

        res = @planner.plan_stream(ctx, goal) do |ev|
          yield(ev) if block_given?
        end
        @memory.recorder.record_plan(res)
        res
      end

      def record_user_input(input)
        @last_user_event_id = @memory.recorder.record_user(input)
        @current_job.add_event(@last_user_event_id) if @current_job
        @last_user_event_id
      end

      def run_call(call)
        tool = call["tool"]
        args = call["args"] || {}
        summary = call["summary"]

        emit(:tool_start, { tool: tool, args: args, summary: summary })

        unless @hooks.run(:before_tool_execution, tool, args)
          emit(:tool_blocked, { tool: tool, reason: "Hook rejected execution" })
          return { status: "blocked", advice: "Execution rejected by hook" }
        end

        st = @validator.status_for(tool)
        unless st[:state] == "ready"
          advice = st[:reason]
          @memory.recorder.record_interception(tool, advice)
          emit(:tool_blocked, { tool: tool, reason: advice })
          return { status: "blocked", advice: advice }
        end
        act = @validator.ensure_active(tool)
        unless act[:ok]
          @memory.recorder.record_interception(tool, act[:advice])
          emit(:tool_blocked, { tool: tool, reason: act[:advice] })
          return { status: "upgrade_required", advice: act[:advice] }
        end

        emit(:tool_executing, { tool: tool })

        res = nil
        modified_files = track_file_modifications do
          res = @engine.execute(tool, args)
        end

        res_payload = { result: res, tool: tool }
        @hooks.run(:after_tool_execution, res_payload)
        res = res_payload[:result]

        if modified_files && !modified_files.empty?
          res = { status: "ok" } unless res.is_a?(Hash)
          res["modified_files"] = modified_files
        end

        emit(:tool_result, { tool: tool, result: res })

        call_seq = @last_user_event_id
        event_id = @memory.recorder.record_execution(tool, res, call_seq: call_seq)
        @current_job.add_event(event_id) if @current_job

        handle_context_lifecycle(tool, args, res)
        begin
          maxc = fetch_call_summary_max
          s = summary.to_s if summary
          s = s[0, maxc] if maxc && s && s.length > maxc
          @memory.recorder.record_summary(s, call_seq || event_id) if s && !s.empty?
        rescue StandardError
        end
        res
      end

      def undo
        @memory.undo
      end

      def redo
        @memory.redo
      end

      private

      def default_memory
        config = Aura::Memory::Config.new(
          store: { project_path: @env_path },
          metabolism: load_config.dig("state_management") || {}
        )
        Aura::Memory::Base.new(
          config: config,
          event_bus: self,
          registry: @registry
        )
      end

      def handle_context_lifecycle(tool, args, res)
        tool_data = @registry.find(tool)
        manifest = tool_data ? (tool_data[:manifest] || {}) : {}

        if manifest["creates_context"]
          ctx_type = manifest["creates_context"]
          ctx_data = res.is_a?(Hash) && res["data"].is_a?(Hash) ? res["data"] : {}
          context_id = res.is_a?(Hash) ? res["context_id"] : nil
          if context_id && !context_id.to_s.empty?
            @context_manager.add_context(ctx_type, ctx_data, id: context_id)
          else
            new_id = @context_manager.add_context(ctx_type, ctx_data)
            res["context_id"] = new_id if res.is_a?(Hash)
          end
        end

        if manifest["destroys_context"]
          destroy_id = args["context_id"] || (res.is_a?(Hash) ? res["context_destroyed"] : nil)
          @context_manager.remove_context(destroy_id) if destroy_id && !destroy_id.to_s.empty?
        elsif manifest["requires_context"]
          use_id = args["context_id"]
          @context_manager.update_activity(use_id) if use_id && !use_id.to_s.empty?
        end
      end

      def fetch_call_summary_max
        load_config.dig("tool_protocol", "call_summary", "max_chars")
      end

      def core_tools_from_config
        list = load_config.dig("tool_protocol", "core_tools") || []
        list.is_a?(Array) ? list.compact.map(&:to_s) : []
      end

      def auto_verify_from_config
        list = load_config.dig("tool_protocol", "auto_verify") || []
        list.is_a?(Array) ? list.compact.map(&:to_s) : []
      end

      def auto_verify_core_tools
        names = (core_tools_from_config + auto_verify_from_config).uniq
        return if names.empty?

        names.each do |name|
          st = @validator.status_for(name)
          next unless st[:state] == "ready"

          @validator.ensure_active(name)
        end
      end

      def track_file_modifications
        before_state = get_file_state
        yield
        after_state = get_file_state

        modified = after_state.keys - before_state.keys

        after_state.each do |path, info|
          next unless before_state[path]

          modified << path if (before_state[path][:mtime] != info[:mtime] || before_state[path][:size] != info[:size]) && !modified.include?(path)
        end

        modified.select! { |f| f.start_with?(@project_path) }
        modified.map! { |f| f.sub(@project_path + "/", "") }
        modified
      end

      def get_file_state
        state = {}
        Dir.glob("#{@project_path}/**/*", File::FNM_DOTMATCH).each do |path|
          next unless File.file?(path)
          next if path.include?("/.git/")
          next if path.include?("/.aura/")

          stat = File.stat(path)
          state[path] = { mtime: stat.mtime.to_i, size: stat.size }
        end
        state
      rescue StandardError
        {}
      end
    end
  end
end

File: src/core/kernel/runner.ts
import { EventEmitter } from 'node:events';
import fs from 'node:fs';
import path from 'node:path';
import { loadTyped } from '../../utils/configManager.js';
import type { AuraConfig } from '../../utils/configSchema.js';
import * as PathResolver from '../../utils/pathResolver.js';
import { ContextAssembler } from '../context/assembler.js';
import { ContextManager } from '../context/manager.js';
import type { ContextPayload } from '../context/payload.js';
import { LSPManager } from '../ext/lsp/manager.js';
import { MemoryBase } from '../memory/base.js';
import { MemoryConfig } from '../memory/config.js';
import { ExecutionEngine } from './executionEngine.js';
import { Hooks } from './hooks.js';
import type {
  IEventBus,
  IRunner,
  PlanEvent,
  PlanResult,
  ToolCall,
  ToolResult,
} from './interfaces.js';
import { Job } from './job.js';
import { Planner } from './planner.js';
import { ToolRegistry } from './registry.js';

export class Runner extends EventEmitter implements IRunner {
  public readonly hooks: Hooks;
  public currentJob: Job | null = null;
  public memory: MemoryBase;
  public planner: Planner;
  public readonly projectPath: string;
  public readonly envPath: string;
  public sessionName: string;

  private registry: ToolRegistry;
  private lspManager: LSPManager;
  private engine: ExecutionEngine;
  private contextManager: ContextManager;
  private configCache: AuraConfig | null = null;
  private lastUserEventId: number | null = null;
  private _autoMode: boolean = false;
  /** Optional abort signal set by the Daemon on socket disconnect. */
  public abortSignal: AbortSignal | null = null;

  public static readonly IGNORED_SCAN_DIRS = [
    '.git',
    '.aura',
    '.aura-workspace',
    'node_modules',
    '.bundle',
    'vendor/bundle',
    'tmp',
    'log',
    'coverage',
    '.next',
    '.nuxt',
    'dist',
    'build',
    '__pycache__',
    '.pytest_cache',
    '.mypy_cache',
    '.venv',
    'venv',
    'env',
    '.cargo',
    'target',
    '.idea',
    '.vscode',
  ];

  constructor(
    projectPath: string,
    options: {
      memory?: MemoryBase;
      registry?: ToolRegistry;
      lspManager?: LSPManager;
      engine?: ExecutionEngine;
      contextManager?: ContextManager;
      hooks?: Hooks;
      planner?: Planner;
    } = {},
  ) {
    super();
    this.projectPath = fs.existsSync(projectPath)
      ? fs.realpathSync(projectPath)
      : path.resolve(projectPath);
    this.envPath =
      PathResolver.environmentPath(this.projectPath) || this.projectPath;
    this.sessionName = process.env.AURA_SESSION_NAME || 'default';

    this.registry = options.registry || new ToolRegistry(this.envPath);
    this.memory = options.memory || this.defaultMemory();
    this.lspManager = options.lspManager || new LSPManager(this.projectPath);
    this.engine =
      options.engine ||
      new ExecutionEngine(this.projectPath, {
        envPath: this.envPath,
        lsp_manager: this.lspManager,
      });
    this.contextManager =
      options.contextManager || new ContextManager(this.envPath);
    this.hooks = options.hooks || new Hooks();
    this.planner =
      options.planner ||
      new Planner(this.projectPath, { envPath: this.envPath });
  }

  public getRegistry(): ToolRegistry {
    return this.registry;
  }

  public getMemory(): MemoryBase {
    return this.memory;
  }

  public getLSPManager(): LSPManager {
    return this.lspManager;
  }

  public getEngine(): ExecutionEngine {
    return this.engine;
  }

  public get workspacePath(): string {
    return this.projectPath;
  }

  public reconnectSession(sessionName: string): void {
    this.sessionName = sessionName;
    process.env.AURA_SESSION_NAME = sessionName;
    if (this.memory?.store) {
      try {
        this.memory.store.close();
      } catch (_e) {}
    }
    this.memory = this.defaultMemory(sessionName);
    this.planner = new Planner(this.projectPath, { envPath: this.envPath });
    this.configCache = null;
  }

  public loadConfig(): AuraConfig {
    if (!this.configCache) {
      this.configCache = loadTyped(this.envPath);
    }
    return this.configCache;
  }

  public clearConfigCache(): void {
    this.configCache = null;
  }

  public startJob(metadata: Record<string, unknown> = {}): Job {
    if (this.currentJob && this.currentJob.status === 'running') {
      throw new Error(`Runner is busy with job ${this.currentJob.id}`);
    }
    this.currentJob = new Job(metadata);
    this.currentJob.start();
    this.emit('job_start', this.currentJob.toObject());
    return this.currentJob;
  }

  public endJob(
    status: 'completed' | 'failed' = 'completed',
    error?: Error | string | null,
  ): Job | null {
    if (!this.currentJob) {
      return null;
    }
    if (status === 'failed' && error) {
      this.currentJob.fail(error);
    } else {
      this.currentJob.complete();
    }
    const job = this.currentJob;
    this.currentJob = null;
    this.emit('job_end', job.toObject());
    return job;
  }

  public async observe(): Promise<ContextPayload> {
    this.memory.recorder.recordCustom('observe', {});
    await this.memory.metabolizeIfNeeded();
    return ContextAssembler.assemble(this.projectPath, this.memory.store.db, {
      lsp_manager: this.lspManager,
    });
  }

  public async plan(
    goal?: string | null,
    context?: unknown,
  ): Promise<PlanResult> {
    const ctx = context || (await this.observe()).toMarkdown();

    const payload: { context: unknown; goal?: string | null } = {
      context: ctx,
      goal,
    };
    await this.hooks.run('before_planning', payload);

    const res = (await this.planner.plan(
      payload.context as string | ContextPayload,
      payload.goal,
    )) as PlanResult;
    this.memory.recorder.recordPlan(res);
    return res;
  }

  public async planStream(
    goal?: string | null,
    context?: unknown,
    onEvent?: (ev: PlanEvent) => void,
  ): Promise<PlanResult> {
    const ctx = context || (await this.observe()).toMarkdown();

    const payload: { context: unknown; goal?: string | null } = {
      context: ctx,
      goal,
    };
    await this.hooks.run('before_planning', payload);

    const res = (await this.planner.planStream(
      payload.context as string | ContextPayload,
      payload.goal || null,
      onEvent,
    )) as PlanResult;
    this.memory.recorder.recordPlan(res);
    return res;
  }

  public recordUserInput(input: string): number {
    this.lastUserEventId = this.memory.recorder.recordUser(input);
    if (this.currentJob) {
      this.currentJob.addEvent(this.lastUserEventId);
    }
    return this.lastUserEventId;
  }

  public async runCall(call: ToolCall): Promise<ToolResult> {
    const tool = call.tool;
    const args = call.args || {};
    const summary = call.summary;

    this.emit('tool_start', { tool, args, summary });

    if (!(await this.hooks.run('before_tool_execution', tool, args))) {
      this.emit('tool_blocked', { tool, reason: 'Hook rejected execution' });
      return { status: 'blocked', advice: 'Execution rejected by hook' };
    }

    this.emit('tool_executing', { tool });

    let res: ToolResult = { status: 'ok' };
    const dbPath = this.memory.store?.dbPath;
    const modifiedFiles = await this.trackFileModifications(async () => {
      // For sleep_and_wake: inject the current abort signal so disconnects
      // cancel the sleep immediately rather than blocking for the full duration.
      const effectiveArgs: Record<string, unknown> =
        tool === 'sleep_and_wake' && this.abortSignal
          ? { ...args, __abortSignal__: this.abortSignal }
          : args;
      res = (await this.engine.execute(tool, effectiveArgs, {
        sessionDbPath: dbPath,
        sessionName: this.sessionName,
      })) as ToolResult;
    });

    const resPayload = { result: res, tool };
    await this.hooks.run('after_tool_execution', resPayload);
    res = resPayload.result as ToolResult;

    if (modifiedFiles && modifiedFiles.length > 0) {
      res.modified_files = modifiedFiles;
    }

    this.emit('tool_result', { tool, result: res });

    const callSeq = this.lastUserEventId;
    const eventId = this.memory.recorder.recordExecution(tool, res, callSeq);
    if (this.currentJob) {
      this.currentJob.addEvent(eventId);
    }

    this.handleContextLifecycle(tool, args, res);

    try {
      const maxc = this.fetchCallSummaryMax();
      let s = summary ? String(summary) : '';
      if (maxc && s.length > maxc) {
        s = s.substring(0, maxc);
      }
      if (s) {
        this.memory.recorder.recordSummary(s, callSeq || eventId);
      }
    } catch (_e) {}

    return res;
  }

  public undo(): boolean {
    return this.memory.undo();
  }

  public redo(): boolean {
    return this.memory.redo();
  }

  public toggleAuto(on: boolean): void {
    this._autoMode = on;
  }

  public get autoMode(): boolean {
    return this._autoMode;
  }

  private defaultMemory(sessionName?: string): MemoryBase {
    const cfg = this.loadConfig();
    const config = new MemoryConfig({
      store: {
        project_path: this.envPath,
        db_path: PathResolver.sessionDbPath(
          this.projectPath,
          sessionName || this.sessionName,
        ),
      },
      metabolism: cfg.state_management ?? {},
    });
    return new MemoryBase({
      config,
      eventBus: this as unknown as IEventBus,
      registry: this.registry,
    });
  }

  private handleContextLifecycle(
    tool: string,
    args: Record<string, unknown>,
    res: ToolResult,
  ): void {
    const toolData = this.registry.find(tool);
    const manifest = toolData ? toolData.manifest || {} : {};

    if (manifest.creates_context) {
      const ctxType = manifest.creates_context;
      const ctxData =
        res && typeof res === 'object' && 'data' in res && res.data
          ? (res.data as Record<string, unknown>)
          : {};
      const contextId =
        res && typeof res === 'object' && 'context_id' in res
          ? (res.context_id as string | null)
          : null;
      if (contextId && String(contextId).trim()) {
        this.contextManager.addContext(ctxType, ctxData, String(contextId));
      } else {
        const newId = this.contextManager.addContext(ctxType, ctxData);
        if (res && typeof res === 'object') {
          (res as Record<string, unknown>).context_id = newId;
        }
      }
    }

    if (manifest.destroys_context) {
      const destroyId =
        (args.context_id as string | undefined) ||
        (res && typeof res === 'object' && 'context_destroyed' in res
          ? (res.context_destroyed as string | null)
          : null);
      if (destroyId && String(destroyId).trim()) {
        this.contextManager.removeContext(String(destroyId));
      }
    } else if (manifest.requires_context) {
      const useId = args.context_id as string | undefined;
      if (useId && String(useId).trim()) {
        this.contextManager.updateActivity(String(useId));
      }
    }
  }

  private fetchCallSummaryMax(): number | null {
    const limit = this.loadConfig().tool_protocol?.call_summary?.max_chars;
    return typeof limit === 'number' ? limit : null;
  }

  private async trackFileModifications(
    fn: () => Promise<void>,
  ): Promise<string[]> {
    const beforeState = this.getFileState();
    await fn();
    const afterState = this.getFileState();

    const modified: string[] = [];
    const _beforeKeys = Object.keys(beforeState);
    const afterKeys = Object.keys(afterState);

    // Added files
    for (const key of afterKeys) {
      if (!beforeState[key]) {
        modified.push(key);
      } else {
        const b = beforeState[key];
        const a = afterState[key];
        if (b.mtime !== a.mtime || b.size !== a.size) {
          modified.push(key);
        }
      }
    }

    return modified
      .filter((f) => f.startsWith(this.projectPath))
      .map((f) => path.relative(this.projectPath, f).replace(/\\/g, '/'));
  }

  private getFileState(): Record<string, { mtime: number; size: number }> {
    const state: Record<string, { mtime: number; size: number }> = {};
    const walk = (dir: string) => {
      let children: string[] = [];
      try {
        children = fs.readdirSync(dir);
      } catch (_e) {
        return;
      }
      for (const name of children) {
        const fullPath = path.join(dir, name);
        try {
          const stat = fs.statSync(fullPath);
          const relative = path
            .relative(this.projectPath, fullPath)
            .replace(/\\/g, '/');

          // Skip ignored directories
          const isIgnored = Runner.IGNORED_SCAN_DIRS.some(
            (d) =>
              relative === d ||
              relative.startsWith(`${d}/`) ||
              relative.includes(`/${d}/`),
          );
          if (isIgnored) continue;

          if (stat.isDirectory()) {
            walk(fullPath);
          } else if (stat.isFile()) {
            state[fullPath] = {
              mtime: Math.floor(stat.mtimeMs),
              size: stat.size,
            };
          }
        } catch (_e) {}
      }
    };
    walk(this.projectPath);
    return state;
  }
}
